src/librustdoc/html/markdown.rs

```diff
@@ -1101,56 +1094,49 @@ fn markdown_summary_with_limit(
     };
 
     let p = Parser::new_with_broken_link_callback(md, opts(), Some(&mut replacer));
-    let p = LinkReplacer::new(p, link_names);
+    let mut p = LinkReplacer::new(p, link_names);
 
-    'outer: for event in p {
+    // FIXME: capacity
```

Comment:
This FIXME is here because the old implementation used `String::with_capacity`. I could add a `HtmlWithLimit::with_capacity` function, but then it could be confusing which parameter is the length limit and which one is the capacity.

Also, are we sure that the capacity estimate is actually useful? We could be wasting memory by allocating too much...

Reply:
I guess one way of getting around the potential confusion of `with_capacity` is to add a `with_buffer` function that allows the user to initialize the buffer themselves.

Reply:
We could simply allocate it with the limit as size.

Reply:
That's a good idea! However, what if the docs are quite short, like `See [`Foo`].`? Then it would waste memory... Maybe it would be best to just add `with_buffer` for now.

Reply:
I think it's even more of a waste to have to re-allocate a few times rather than allocate once but too much. Memory works in pages and realloc only move the value if there isn't enough space after the end of your current pointer. So all in all, I think it'd be a wiser move to allocate "too much" by default.

Reply:
Maybe @jyn514 has an opinion on this point?

Reply:
> I think it's even more of a waste to have to re-allocate a few times rather than allocate once but too much. Memory works in pages and realloc only move the value if there isn't enough space after the end of your current pointer. So all in all, I think it'd be a wiser move to allocate "too much" by default.

I tested your suggestion of using the limit as the initial capacity to see how much memory it would waste. Unfortunately, the results don't look great. When documenting `std` (and `core`, etc.), we allocate about half a megabyte (~410 kB to be more precise) more than is necessary. This is assuming that the rendered summary string is not modified after `HtmlWithLimit::finish()`; the summary string *is* sometimes modified later, when the `…` is sometimes added at the end. However, I think those extra few bytes that are occasionally added are unlikely to have a huge effect on the results.

As a percentage, rustdoc ends up allocating 34% more memory than is necessary.

Here are the raw numbers:

|||
|-|-|
| **Total number of summary string allocations** | 23538 |
| **Total bytes needed (i.e., sum of the lengths)** | 1227676 |
| **Total bytes allocated (i.e., sum of the capacities)** | 1646218 |

Reply:
Thanks for the numbers! Do you have the equivalent with `String::new` by any chance?

Reply:
No, but I can measure that.

Reply:
Yes please. :)

Reply:
Wow, good catch! The numbers are almost identical. So much so that I'm suspicious I'm measuring incorrectly, but there are slight differences:

Total # of allocs: 23538
Total length: 1227676 bytes
Total capacity: 1646218 bytes
Wasted memory: 418542 bytes (34% more than necessary)

Reply:
Actually, there aren't any differences. I must have measured incorrectly...

Reply:
So using `with_capacity` with the length limit is actually *more* memory efficient 😆

Reply:
I don't think this is worth worrying about, I would just use `String::new` unless this comes up in a profiler. `collect_blanket_impls` will still be 100x as slow either way :P

Reply:
Maybe, but it's not a reason to not pay attention on the rest. ;)

Reply:
Yeah, it's probably not a big deal, but since we were preallocating before, I figure it's a good idea to try to keep the previous allocation behavior :)